### PR TITLE
Enhance UPRN and fix loading for Scotland data

### LINKED ISSUE #72: Create unique identifier for properties
- [x] For some propeties, we have different EPC entries for the same property sometimes with a UPRN value, other times with missing UPRN. We should fill the missing with the existing UPRN;
- [x] Create a unique identifier that enables uniquely identifying properties even those with missing UPRN.

### LINKED ISSUE #86: Changes to Scotland raw field names, encoding and column rows

### PR DESCRIPTION
**Pull request by sofiapinto**
# Description 

This PR fixes:
- #72: prior to this PR, missing `UPRN` was filled with the `BUILDING_REFERENCE_NUMBER` but only for England and Wales, hence `UPRN` had missing values for Scotland. This original enhancement was in script `asf_core_data/getter/epc/epc_data.py`. I removed the enhancement from that script, so that the raw data has missing `UPRN`. We instead now use `BUILDING_REFERENCE_NUMBER` together with address information (`ADDRESS1`, `ADDRESS2` and `POSTCODE`) to enhance `UPRN` during the processing pipeline - changes made to `asf_core_data/pipeline/preprocessing/feature_engineering.py`. Hence, now EPC processed/processed and deduplicated should have no missing `UPRN`.
- #86: I think this has already been fixed by Solomon in asf_daps so no need to change it on your side if loading Scotland data works for you. But in asf_core_data we have problems when loading Scotland data due to column name changes and multiple column headers - so Julia implemented a fix in `load_scotland_data()` in `asf_core_data/getter/epc/epc_data.py` and in `asf_core_data/config/base_config.py`. These changes haven't been committed before, so I'm committing them now.

closes #72
closes #86

# Instructions for Reviewer(s)

## Review

Hey @Jack-Vines and @sqr00t , thanks for reviewing these scripts. Could one of you please double check if the logic looks okay for the changes in `asf_core_data/pipeline/preprocessing/feature_engineering.py`? Thanks a lot!

## Setup
In case you want to run anything (no need to!!!):

- clone this repo: `git clone git@github.com:nestauk/asf_core_data.git`
- checkout to the correct branch: `git checkout 72_fix_property_id_and_epc_loading`
- Run `make install`;
- Run `direnv allow`;
- Activate the conda environment: `conda activate asf_core_data`;
- Run `python asf_core_data/pipeline/preprocessing/preprocess_epc_data.py` (this will load data from S3, so might take some time). If you have raw data in a local folder you can do `python asf_core_data/pipeline/preprocessing/preprocess_epc_data.py --path_to_data "YOUR/LOCAL/PATH/TO/DATA`. Note that, either way, this will only store the processed data locally.
---

### CODE REVIEW

File: asf_core_data/getters/epc/epc_data.py [hunk 1]
```diff
@@ -185,8 +185,8 @@ def load_scotland_data(
             dtype=dtype,
             low_memory=low_memory,
             usecols=scot_usecols,
-            skiprows=skiprows,  # don't load first row (more ellaborate feature names),
-            encoding="latin-1",
```
Comment (sqr00t) [inline asf_core_data/getters/epc/epc_data.py:189]: Finally! Ridding of latin-1 encoding was much anticipated!

File: asf_core_data/pipeline/preprocessing/feature_engineering.py [hunk 2]
```diff
@@ -43,6 +44,101 @@
...
+        building_reference: building reference number, e.g. 1234
+        address_info: address information, e.g. "flat 1 abc road_AB2Y8X"
+
+    Returns:
+        A string with concatenated building reference and address info (e.g. "1234_flat 1 abc road_AB2Y8X")
+    """
+    return str(building_reference) + "_" + address_info
+
+
+def enhance_uprn(df: pd.DataFrame) -> pd.DataFrame:
```
Comment (sqr00t) [inline asf_core_data/pipeline/preprocessing/feature_engineering.py:87]: Overrall, I quite like this PR as the changes are specific, and the logic is well commented.

I would extract parts of this out as it's own function (i.e. mapping creation) on the asf_daps side. But I'm happy to merge this PR here. (oops, not the best quoting of the code)

**Review by Jack-Vines — APPROVED** [review]
looks good to me too!

**Review by sqr00t — APPROVED** [review]

File: asf_core_data/pipeline/preprocessing/feature_engineering.py (hunk 2 above)
Comment (sofiapinto) [reply asf_core_data/pipeline/preprocessing/feature_engineering.py:87]: thanks for review @sqr00t ! Agree with further refactoring the `enhance_uprn()` function, but if you'll do it on asf_daps side, i'll leave it like this there.

**Pull request merged**